### ISSUE TITLE
Purpose of ast.Statement class?

### ISSUE THREAD

**Issue by jlumpe** [bjodah/sympy#10]
I'm trying to understand the hierarchy of ast classes here. As far as I can see Statement is just a wrapper for a single argument, but doesn't seem to be used anywhere. It seems like Statement should be the base class for most other classes like Assignment, For, While, Return, Break etc. A CodeBlock or function body would then correspond to a sequence of Statement instances.

**Comment by bjodah** [comment]
It was introduced because some nodes, when printed as C code, sometimes needed a semicolon, and sometimes not. (e.g. a function call). greping for `Statement` (to refresh my own memory):
```
$ grep Statement -R .
./core/tests/test_args.py:def test_sympy__codegen__ast__Statement():
./core/tests/test_args.py:    from sympy.codegen.ast import Statement, AddAugmentedAssignment
./core/tests/test_args.py:    assert _test_args(Statement(AddAugmentedAssignment(x, -1)))
./printing/codeprinter.py:    def _print_Statement(self, expr, **kwargs):
./printing/tests/test_ccode.py:    While, Scope, Print, FunctionPrototype, FunctionDefinition, FunctionCall, Statement, Return,
./printing/tests/test_ccode.py:    assert ccode(Statement(x)) == 'x;'
./printing/tests/test_ccode.py:    assert ccode(Statement(Print([x, y], "%d %d"))) == 'printf("%d %d", x, y);'
./printing/tests/test_ccode.py:    assert ccode(Statement(FunctionCall('pwer', [x]))) == 'pwer(x);'
./printing/tests/test_ccode.py:    assert ccode(Statement(Return(x))) == 'return x;'
./codegen/fnodes.py:    Attribute, CodeBlock, Declaration, FunctionCall, Node, none, Statement, String,
./codegen/algorithms.py:    Print, Return, Scope, Statement, While, Variable, Pointer, real
./codegen/algorithms.py:        prnt = Statement(Print([wrt, delta], r"{0}=%12.5g {1}=%12.5g\n".format(wrt.name, name_d)))
./codegen/algorithms.py:    declars = [Statement(Declaration(Variable(delta, type=real, value=oo)))]
./codegen/algorithms.py:    body = CodeBlock(algo, Statement(Return(wrt)))
./codegen/tests/test_applications.py:    integer, Variable, CodeBlock, While, Statement
./codegen/tests/test_applications.py:    whl = While(i<n, [Assignment(out[i], inp[i]), Statement(PreIncrement(i))])
./codegen/tests/test_applications.py:    body = CodeBlock(Statement(i.as_Declaration(value=0)), whl)
./codegen/ast.py:       |--->Statement
./codegen/ast.py:class Statement(Basic):
./codegen/ast.py:    >>> from sympy.codegen.ast import FunctionDefinition, Return, Statement
./codegen/ast.py:    >>> body = [Statement(Return(x*y))]
./codegen/cnodes.py:from sympy.codegen.ast import Attribute, Declaration, Node, Statement, String, Token, Type, none, FunctionCall
```
It should probably not be imported in `.codegen.fnodes` at least.

If we could do without `Statement` I am all for it (i.e. making the test suite pass without it). It doesn't feel very language agnostic, but since e.g. calling functions in C can be done both as a stand-alone statement and as part of an expression we will probably need some equivalent mechanism. What do you think?

**Comment by jlumpe** [comment]
I think the statement class definitely could be useful because it's a fairly fundamental programming concept that applies to most languages. My idea for the class hierarchy would be something like this:

* Basic (or AST base class that subclasses Basic - might be useful for printers to know they're printing code)
    * Type
    * Expression
        * Variable
        * String (or StringLiteral?)
    * Statement
        * Assignment
            * AugmentedAssignment...
        * Declaration
        * If (not yet implemented?)
        * For
        * While
        * Print
        * Break
        * Continue
        * Return
        * FunctionCall
        * FunctionPrototype
            * FunctionDefinition
        * Comment

A module or function body could then be easily defined to be a sequence of statements. Each statement takes up its own line or several lines. That should definitely apply to C, Python, and a lot of others, and I think Fortran. The C printer would probably be on a case-by-case basis but in general the single-line statements (Declaration, Assignment, FunctionCall, Break, Continue, Return) would have a semicolon.

**Comment by bjodah** [comment]
Sounds good. I more or less made `Token` the base class for the new additions to the `.codegen.ast` module, but I didn't change the previously existing nodes to use it (I would always break tests -- and hence backward compatibilty -- whenever I tried). If you run into the same problem I would recommend that we just add new versions with other names and deprecate the old ones (we would need to show that the new ones have some benefits though), that way we can have a proper deprecation cycle.

I agree with your re-ordering of the class hierarchy (having `FunctionDefinition` under `Statement` feels a bit odd but I guess it makes sense from the printers point of view). That said, there are often corner cases that appear when trying to implement things. I found that it was helpful to have an application in mind when I wrote the classes (the most extensive one was Newton's method). I guess you have some applications in mind yourself? The more use cases we can cover with our nodes, the more confidence we'll have that this is the right approach.
